### ISSUE TITLE
FIX `phpenv versions` should not include partial builds

### ISSUE THREAD

**Issue by allanlaal** [phpenv/phpenv#115]
I had a lot of `phpenv install reallyoldversion`s fail (because it needs old openssl: https://github.com/php-build/php-build/issues/718), but php-build does not clean up failed builds https://github.com/php-build/php-build/issues/716

only 8.1.16 has built successfully:

	  allan@L4:~$ phpenv versions
	  * system (set by /home/allan/.phpenv/version)
	    5.3.226
	    5.3.29
	    5.6.39
	    5.6.40
	    7.1
	    7.1.20
	    7.1.30
	    7.1.32
	    7.1.33
	    7.1snapshot
	    7.4
	    7.4.29
	    8.0.19
	    8.1
	    8.1.0
	    8.1.6

(note that the versions without a minor version are symlinks generated by [phpenv alias plugin](https://github.com/madumlao/phpenv-aliases))

a simple check for versions list would be to check if bin/php is present for that version

**Comment by debo** [comment]
As far as I can tell this is an intended behaviour. I know it's annoying but I don't see this as a big deal. @madumlao what are you thoughts on this one?

**Comment by madumlao** [comment]
The cleanup issue might be a php-build issue, not phpenv.

All the *env build systems allow for you to put any arbitrary folder or link in your `versions` directory, because you can technically manually compile PHP yourself and link the results into versions, without the equivalent of php-build.

Actually checking that the installed prefix is valid through an executable `bin/php` might be a good idea though and is probably something to suggest to all upstreams. However, the question would be what should be the expected behavior? Do we just stop at an executable `bin/php` file, or does a deeper check make sense? And given that the check may run with every invocation of every executable (php shim will prefix every command in your path) you want this check to be very fast and have no weird behavior that might break a user's shell.

**Comment by debo** [comment]
Indeed, personally, I feel that this functionality is a nice to have which will have little added benefit to someone workflow at a cost of adding way too much complexity in the tool and likely performances issues too. I would pass.

**Comment by allanlaal** [comment]
performance is a good point, its better if this is fixed upstream https://github.com/php-build/php-build/issues/716

**Comment by madumlao** [comment]
phpenv versions will only list directories in the versions directory. There is no checking done and no known way to confirm if the directory is truly a valid PHP installation. You can also symlink directories there to PHP installations elsewhere on the system.

This bug belongs in php-build, and has been reported there, hence wiill be closed here. php-build should ideally not populate the versions directory until the installation is successful, which I think is in line with how the other env/build projects like ruby-build, works.

**Comment by debo** [comment]
Maybe something for the future @madumlao but thinking about this in hindsight, wouldn't be sufficient to check error codes and status codes at the end of `phpenv` execution? Or once we invoke `php-build` it's out of our control? I didn't check the code, simply thinking about a possible way.

**Comment by debo** [comment]
Unless we, for example, add a lock file when the build is completed successfully which we can trap. But I agree possibly this is more of a Could Have then a Must/Should

**Comment by madumlao** [comment]
Reopening to track issue in php-build
https://github.com/php-build/php-build/issues/716

> OR you need to modify phpenv-install to clear out the directory on failed install.

On further review, this is what upstream [ruby-build](https://github.com/rbenv/ruby-build/blob/master/bin/rbenv-install) does. I also see cleanup code in php-build, but it is either not being called or not running in phpenv-install.

As this is apparently an issue specifically in integration of phpenv and php-build, I will track in both places.
